### PR TITLE
Fix parsing of default values in `buildASTSchema`, `extendSchema` and `buildClientSchema`

### PR DESCRIPTION
**Pull request by IvanGoncharov**
In the current implementation `buildASTSchema`, `extendSchema` and `buildClientSchema` register following stubs as `parseLiteral` and `parseValue`:
```js
    return new GraphQLScalarType({
      // ...

      // Note: validation calls the parse functions to determine if a
      // literal value is correct. Returning null would cause use of custom
      // scalars to always fail validation. Returning false causes them to
      // always pass validation.
      parseValue: () => false,
      parseLiteral: () => false,
    });
```

Because of that, `graphql-js` can't correctly parse default values of custom scalar types:
```graphql
scalar CustomScalar

type Query {
  someField(someArg: CustomScalar = 5): String
}
```

So to clarify. You can override those functions after the schema is built, but the default value from IDL will be lost.

To illustrate this I've added a test case in the first commit of this PR and here is the result:
screen shot 2017-06-09 at 11 20 21 pm

This PR solves this problem by providing reasonable defaults for `parseLiteral` and `parseValue`.

### CODE REVIEW

File: src/utilities/scalarValueFromAST.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+import keyValMap from '../jsutils/keyValMap';
+import * as Kind from '../language/kinds';
+import type { ValueNode } from '../language/ast';
+
+/**
+ * Create a JavaScript value from a GraphQL languate AST representation
+ * of Scalar.
+ */
+export function scalarValueFromAST(astValue: ValueNode): mixed {
```
Comment (leebyron) [inline src/utilities/scalarValueFromAST.js]: This looks very similar to `valueFromAST` - can that be used instead?

File: src/type/definition.js [hunk 2]
```diff
@@ -333,7 +334,7 @@ export class GraphQLScalarType {
   // Parses an externally provided value to use as an input.
   parseValue(value: mixed): mixed {
     const parser = this._scalarConfig.parseValue;
-    return parser && !isNullish(value) ? parser(value) : undefined;
+    return parser && !isNullish(value) ? parser(value) : value;
```
Comment (leebyron) [inline src/type/definition.js]: If a value parses as nullish, shouldn't it return clearly as `undefined` to not be confused with a true `null` value?

File: src/utilities/index.js [hunk 3]
```diff
@@ -58,6 +58,10 @@ export { valueFromAST } from './valueFromAST';
 // Create a GraphQL language AST from a JavaScript value.
 export { astFromValue } from './astFromValue';
 
+// Create a JavaScript value from a GraphQL languate AST representation
```
Comment (leebyron) [inline src/utilities/index.js]: language

File: src/utilities/scalarValueFromAST.js (hunk 1 above)
Comment (leebyron) [inline src/utilities/scalarValueFromAST.js]: If this function is for scalars, then it should likely not support LIST or OBJECT, correct?

Comment (IvanGoncharov) [inline src/utilities/scalarValueFromAST.js]: > This looks very similar to valueFromAST - can that be used instead?

`valueFromAST` builds value based on the type definition it accepts as the second argument.
In a case of a scalar, there is no such type definition.

> If this function is for scalars, then it should likely not support LIST or OBJECT, correct?

AFAIK, the spec doesn't explicitly forbid complex scalars as LIST or OBJECT. Moreover, it's a widespread practice to create such scalars. More details in this PR:
https://github.com/facebook/graphql/pull/325

File: src/utilities/index.js (hunk 3 above)
Comment (IvanGoncharov) [inline src/utilities/index.js]: Fixed

File: src/type/definition.js (hunk 2 above)
Comment (IvanGoncharov) [inline src/type/definition.js]: Fixed.

**Comment by leebyron** [comment]
Thanks for making this happen, Ivan. Great improvement

**Pull request closed without merging**